src/util/syncbuffer.cpp

```diff
@@ -37,7 +37,13 @@ SyncBuffer::~SyncBuffer() {} // NOLINT
 void SyncBuffer::encode(const std::string& s) {
     ghoul_assert(_encodeOffset + sizeof(char) * s.size() + sizeof(int32_t) < _n, "");
 
-    int32_t length = static_cast<int32_t>(s.length());
+    int32_t anticpatedBufferSize = _encodeOffset + (sizeof(char) * s.size())
+        + sizeof(int32_t);
+    if (anticpatedBufferSize >= _n) {
+        _dataStream.resize(anticpatedBufferSize);
+    }
+
+    int32_t length = static_cast<int32_t>(s.size() * sizeof(char));
     memcpy(
```

Comment:
If you change the memcpy to a std::insert, or similar (https://en.cppreference.com/w/cpp/container/vector/insert), you don't need to explicitly resize the vector before. I think that would be easier to read and less prone to errors.

Reply:
I had originally considered doing this, but decided not to make this change. This code is using a `std::vector`, but accesses its memory directly without using any of the safe/convenient features of the container (besides resize). `memcpy` is very fast, so I suspect this was written for performance, and so I'll keep it as-is. Otherwise I would want to rewrite everything to use the vector member functions for access.

Reply:
Check this for performance comparison between insert and memcpy:
http://thewayofc.blogspot.com/2014/05/copying-memory-from-c-to-c-using.html?m=1

Reply:
When I say that `memcpy` is fast, I'm not advocating using it over C++ standard operators. I'm just saying this is the only reason I can think of why it had been written that way.
The options I see for this fix:
1. Stick with `memcpy` to stay consistent with the "C style" of memory access that this file currently uses (I'm not counting resize because it was already in there).
2. Use a mix of C and C++ styles of memory access by introducing std::insert.
3. Rewrite everything to use C++ standard operators.

If I was writing this from scratch I would use only C++ standard operations on a vector. Even if there was a performance hit, it would be more readable. I have other things to work on right now so option #3 is out, at least for now.

Functionally, either way will work but what is best for code readability? For this fix, I prefer option 1 because the code is more consistent. I think that mixing access types makes it more confusing to the next person. What is your opinion on option 1 or 2? I will make my change based on which you think is best for readability.